File: app/src/main/java/com/twisthenry8gmail/fullcart/UseByReceiver.java
package com.twisthenry8gmail.fullcart;

import android.app.PendingIntent;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;

import java.util.ArrayList;

import androidx.core.app.NotificationCompat;
import androidx.core.app.NotificationManagerCompat;
import androidx.core.content.ContextCompat;

public class UseByReceiver extends BroadcastReceiver {

    @Override
    public void onReceive(Context context, Intent intent) {

        showNotification(context, 3, true);
    }

    static void showNotification(Context context, int maxItems, boolean autoCancel) {

        DatabaseHelper databaseHelper = DatabaseHelper.getInstance(context);
        String[] columns = {DatabaseHelper.LIST_NAME, DatabaseHelper.LIST_CATEGORY, DatabaseHelper.PANTRY_BY_DATE};
        String selection = "length(" + DatabaseHelper.PANTRY_BY_DATE + ")!=1";
        Cursor cursor = databaseHelper.getReadableDatabase().query(DatabaseHelper.PANTRY_TABLE_NAME, columns, selection, null, null, null, DatabaseHelper.PANTRY_BY_DATE);

        ArrayList<String> items = new ArrayList<>();

        int diffPreference = SettingsFragment.getPreferredDateDifference(context);

        while (cursor.moveToNext()) {
            String byDate = cursor.getString(cursor.getColumnIndex(DatabaseHelper.PANTRY_BY_DATE)).substring(1);

            long diff = DateUtil.getCurrentSimpleDateDifference(byDate);

            if (diff <= diffPreference) {

                String name = cursor.getString(cursor.getColumnIndex(DatabaseHelper.LIST_NAME));
                String daysLeftString;

                if (diff == 0) {
                    daysLeftString = context.getString(R.string.notification_use_by_content_pattern_present, name);
                }
                else {
                    String unit = Math.abs(diff) == 1 ? context.getString(R.string.day) : context.getString(R.string.days);
                    if (diff < 0) {
                        daysLeftString = context.getString(R.string.notification_use_by_content_pattern_past, name, -diff, unit);
                    }
                    else {
                        daysLeftString = context.getString(R.string.notification_use_by_content_pattern, name, diff, unit);
                    }
                }

                items.add(daysLeftString);

                if (items.size() == maxItems) {
                    break;
                }
            }
        }
        cursor.close();

        if (items.size() > 0) {

            Intent contentIntent = new Intent(context, MainActivity.class);
            contentIntent.putExtra(MainActivity.SHOW_PANTRY, true);
            PendingIntent contentPendingIntent = PendingIntent.getActivity(context, 0, contentIntent, 0);

            NotificationCompat.Builder builder = new NotificationCompat.Builder(context, AlarmUtil.CHANNEL_USE_BY);

            if (items.size() > 1) {

                NotificationCompat.InboxStyle style = new NotificationCompat.InboxStyle();
                for (String item : items) {
                    style.addLine(item);
                }

                builder.setStyle(style);
            }

            builder.setSmallIcon(R.drawable.notification_icon)
                    .setColor(ContextCompat.getColor(context, R.color.primary))
                    .setContentTitle(context.getString(R.string.notification_use_by_title))
                    .setContentText(items.get(0))
                    .setContentIntent(contentPendingIntent).setAutoCancel(autoCancel);

            NotificationManagerCompat.from(context).notify(0, builder.build());
        }
    }
}

File: app/src/main/java/uk/henrytwist/fullcart/services/UseByReceiver.kt
package uk.henrytwist.fullcart.services

import android.app.Notification
import android.app.PendingIntent
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import uk.henrytwist.fullcart.R
import uk.henrytwist.fullcart.models.PantryNotification
import uk.henrytwist.fullcart.usecases.GetPantryNotifications
import uk.henrytwist.fullcart.util.NotificationUtil
import uk.henrytwist.fullcart.view.ListItemFormatter
import uk.henrytwist.fullcart.view.main.MainActivity
import javax.inject.Inject

@AndroidEntryPoint
class UseByReceiver : BroadcastReceiver() {

    @Inject
    lateinit var getPantryNotifications: GetPantryNotifications

    override fun onReceive(context: Context?, intent: Intent?) {

        context ?: return

        GlobalScope.launch(Dispatchers.IO) {

            val notificationData = getPantryNotifications()

            if (notificationData.isNotEmpty()) {

                val notificationManager = NotificationManagerCompat.from(context)

                notificationData.forEach {

                    val notification = buildNotification(context, it)
                    notificationManager.notify(BASE_NOTIFICATION_ID + 1 + it.listId, notification)
                }

                notificationManager.notify(BASE_NOTIFICATION_ID, buildSummaryNotification(context, notificationData))
            }
        }
    }

    private fun buildNotification(context: Context, data: PantryNotification): Notification {

        val intent = Intent(context, MainActivity::class.java)
        intent.putExtra(MainActivity.FROM_NOTIFICATION_LIST_ID, data.listId)
        val pendingIntent = PendingIntent.getActivity(context, data.listId, intent, 0)
        return NotificationCompat.Builder(context, NotificationUtil.CHANNEL_USE_BY)
                .setSmallIcon(R.drawable.notification_icon)
                .setContentTitle(data.listName)
                .setStyle(buildStyle(context, data))
                .setGroup(GROUP_KEY)
                .setContentText(buildContentText(data))
                .setContentIntent(pendingIntent)
                .build()
    }

    private fun buildStyle(context: Context, data: PantryNotification): NotificationCompat.Style {

        val style = NotificationCompat.InboxStyle()

        data.items.forEach {

            val useByFormatted = ListItemFormatter.getUseByDateSummary(context, it.useByDate)!!
            style.addLine(context.getString(R.string.use_by_notification_item, it.name, useByFormatted))
        }

        return style
    }

    private fun buildContentText(data: PantryNotification): String {

        return data.items.joinToString(" · ") { it.name }
    }

    private fun buildSummaryNotification(context: Context, data: List<PantryNotification>): Notification {

        val intent = Intent(context, MainActivity::class.java)
        val pendingIntent = PendingIntent.getActivity(context, 0, intent, 0)
        return NotificationCompat.Builder(context, NotificationUtil.CHANNEL_USE_BY)
                .setSmallIcon(R.drawable.notification_icon)
                .setSubText(context.getString(R.string.use_by_notification_title))
                .setGroup(GROUP_KEY)
                .setStyle(buildSummaryStyle(data))
                .setContentIntent(pendingIntent)
                .setGroupSummary(true)
                .build()
    }

    private fun buildSummaryStyle(data: List<PantryNotification>): NotificationCompat.Style {

        val style = NotificationCompat.InboxStyle()

        data.forEach {

            style.addLine(it.listName)
        }

        return style
    }

    companion object {

        const val BASE_NOTIFICATION_ID = 0
        const val GROUP_KEY = "uk.henrytwist.fullcart.USE_BY_GROUP"
    }
}